pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,195 @@
+		Type:   v1alpha1.CertificateRequestConditionReady,
+		Status: v1alpha1.ConditionFalse,
+		Reason: errorCertificateFailed,
+	}) {
+		dbg.Info("certificate request condition failed so skipping processing")
+		return nil
+	}
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("fetching issuer object referenced by CertificateRequest")
+
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		log.WithValues(
+			logf.RelatedResourceNameKey, crCopy.Spec.IssuerRef.Name,
+			logf.RelatedResourceKindKey, crCopy.Spec.IssuerRef.Kind,
+		).Error(err, "failed to find referenced issuer")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	dbg.Info("ensuring issuer type matches this controller")
+
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	issuerType, err := apiutil.NameForIssuer(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		log.Error(err, "failed to obtain referenced issuer type")
+		return nil
+	}
+
+	// This CertificateRequest is not meant for us, ignore
+	if issuerType != c.issuerType {
+		c.log.WithValues(
+			logf.RelatedResourceKindKey, issuerType,
+		).V(5).Info("issuer reference type does not match controller resource kind, ignoring")
+		return nil
+	}
+
+	dbg.Info("validating CertificateRequest resource object")
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(crCopy.Status.Certificate) > 0 {
+		dbg.Info("certificate field is already set in status so skipping processing")
+		c.setCertificateRequestStatus(crCopy)
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	// TODO: Metrics??
+
+	dbg.Info("invoking sign function as existing certificate does not exist")
+	return c.sign(ctx, crCopy, i)
+}
+
+// return an error on failure. If retrieval is succesful, the certificate data
+// will be stored in the certificate request status
+func (c *Controller) sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuer issuer.Interface) error {
+	log := logf.FromContext(ctx)
+
+	defer c.setCertificateRequestStatus(cr)
+
+	resp, err := issuer.Sign(ctx, cr)
+	if err != nil {
+		log.Error(err, "error issuing certificate request")
+		return err
+	}
+
+	// if the issuer has not returned any data, exit early
+	if resp == nil {
+		return nil
```

Comment:
Sorry for the belated commentary, but the purpose of returning nil here seems unclear.    Would the controller reschedule the call to the issuer?   Would this permit an asynchronous issuer?

Reply:
Hello! :wave: 

Yeah that's right - we want to exit the sync here if the underlying sign function isn't ready yet. We should be synced again when appropriate and try again. This means that the sign function needs to be idempotent.

This is a bit clearer with the WIP acme CR implementation;
https://github.com/jetstack/cert-manager/pull/1943/files#diff-b5082e2ea140a11ff02a0f075decec76R165
Here we re-sync when the status of the order resource changes.